crates/ra_syntax/src/validation.rs

```diff
@@ -111,55 +111,46 @@ pub(crate) fn validate(root: &SyntaxNode) -> Vec<SyntaxError> {
     errors
 }
 
-// FIXME: kill duplication
 fn validate_literal(literal: ast::Literal, acc: &mut Vec<SyntaxError>) {
+    fn unquote(text: &str, prefix_len: usize, end_delimiter: char) -> Option<&str> {
+        text.rfind(end_delimiter).and_then(|end| text.get(prefix_len..end))
+    }
+
     let token = literal.token();
     let text = token.text().as_str();
+
+    let mut push_err = |prefix_len, (off, err): (usize, unescape::EscapeError)| {
+        let off = token.text_range().start() + TextUnit::from_usize(off + prefix_len);
+        acc.push(SyntaxError::new(err.into(), off));
+    };
+
     match token.kind() {
         BYTE => {
-            if let Some(end) = text.rfind('\'') {
-                if let Some(without_quotes) = text.get(2..end) {
-                    if let Err((off, err)) = unescape::unescape_byte(without_quotes) {
-                        let off = token.text_range().start() + TextUnit::from_usize(off + 2);
-                        acc.push(SyntaxError::new(err.into(), off))
-                    }
-                }
+            if let Some(Err(e)) = unquote(text, 2, '\'').map(unescape::unescape_byte) {
+                push_err(2, e);
             }
         }
         CHAR => {
-            if let Some(end) = text.rfind('\'') {
-                if let Some(without_quotes) = text.get(1..end) {
-                    if let Err((off, err)) = unescape::unescape_char(without_quotes) {
-                        let off = token.text_range().start() + TextUnit::from_usize(off + 1);
-                        acc.push(SyntaxError::new(err.into(), off))
-                    }
-                }
+            if let Some(Err(e)) = unquote(text, 1, '\'').map(unescape::unescape_char) {
+                push_err(1, e);
             }
         }
         BYTE_STRING => {
-            if let Some(end) = text.rfind('\"') {
-                if let Some(without_quotes) = text.get(2..end) {
-                    unescape::unescape_byte_str(without_quotes, &mut |range, char| {
-                        if let Err(err) = char {
-                            let off = range.start;
-                            let off = token.text_range().start() + TextUnit::from_usize(off + 2);
-                            acc.push(SyntaxError::new(err.into(), off))
-                        }
-                    })
-                }
+            if let Some(without_quotes) = unquote(text, 2, '"') {
+                unescape::unescape_byte_str(without_quotes, &mut |range, char| {
+                    if let Err(err) = char {
+                        push_err(2, (range.start, err));
+                    }
+                })
```

Comment:
We could do the same here as described for char, e.g. pull into a separate function. I think that would be nice :)

Reply:
I've stumbled with a language limitation here.
Since here we would have a second level higher order function
something like
```rust
fn validate_byte_str_or_str_literal(..., impl FnOnce(&str, impl FnMut(...))) {}
```
Notice the nested `impl FnMut(...)`

Reply:
It would be a big and ugly hack to do it with a dynamic dispatch here.
`unescape::unescape_str(&str, impl FnMut(Range<usize>, Result<u8, EscapeError>))`
This function expected a `Sized` second argument.
This means we would end up with the following hack:
```rust

fn validate_byte_or_char_str_literal<T>(
    token: &SyntaxToken,
    prefix_len: usize,
    suffix_len: usize,
    acc: &mut Vec<SyntaxError>,
    unescape_fn: impl FnOnce(&str, &mut dyn FnMut(Range<usize>, Result<T, unescape::EscapeError>)),
) {
    let text = token.text().as_str();
    if let Some(without_quotes) = unquote(text, prefix_len, suffix_len) {
        unescape_fn(without_quotes, &mut |range, char| {
            if let Err(err) = char {
                let off = token.text_range().start() + TextUnit::from_usize(range.start + prefix_len);
                acc.push(SyntaxError::new(err.into(), off));
            }
        });
    }
}


validate_byte_or_char_str_literal(
    &token,
    2,
    1,
    acc,
    |s, cb| unescape::unescape_byte_str(s, &mut |range, char| cb(range, char)),
);
```

Reply:
I've been unable make the above function reasonable, e.g. less lines of code than what it replaces, so lets just keep it at de-duplicating char and byte for now.

Reply:
Same opinion here 😏 :